components/core/src/clp/ir/LogEventSerializer.cpp

```diff
@@ -0,0 +1,159 @@
+#include "LogEventSerializer.hpp"
+
+#include <string>
+#include <string_view>
+
+#include <spdlog/spdlog.h>
+
+#include "../Defs.h"
+#include "../ErrorCode.hpp"
+#include "../ffi/ir_stream/encoding_methods.hpp"
+#include "../ffi/ir_stream/protocol_constants.hpp"
+#include "../ir/types.hpp"
+#include "../type_utils.hpp"
+
+using std::string;
+using std::string_view;
+
+namespace clp::ir {
+template <typename encoded_variable_t>
+LogEventSerializer<encoded_variable_t>::~LogEventSerializer() {
+    if (m_is_open) {
+        SPDLOG_ERROR("clp::ir::LogEventSerializer not closed before being destroyed - output maybe "
+                     "corrupted.");
+    }
+}
+
+template <typename encoded_variable_t>
+auto LogEventSerializer<encoded_variable_t>::open(string const& file_path) -> bool {
+    if (m_is_open) {
+        throw OperationFailed(ErrorCode_NotReady, __FILENAME__, __LINE__);
+    }
+
+    m_serialized_size = 0;
+    m_num_log_events = 0;
+    m_ir_buf.clear();
+
+    m_writer.open(file_path, FileWriter::OpenMode::CREATE_FOR_WRITING);
+    m_zstd_compressor.open(m_writer);
+
+    bool res{};
+    if constexpr (std::is_same_v<encoded_variable_t, four_byte_encoded_variable_t>) {
+        m_prev_event_timestamp = 0;
+        res = ffi::ir_stream::four_byte_encoding::serialize_preamble(
+                cTimestampPattern,
+                cTimestampPatternSyntax,
+                cTimezoneID,
+                m_prev_event_timestamp,
+                m_ir_buf
+        );
+    } else {
+        res = clp::ffi::ir_stream::eight_byte_encoding::serialize_preamble(
+                cTimestampPattern,
+                cTimestampPatternSyntax,
+                cTimezoneID,
+                m_ir_buf
+        );
+    }
+
+    if (false == res) {
+        close_writer();
+        return true;
+    }
+
+    m_is_open = true;
+
+    // Flush the preamble
+    flush();
+
+    return false;
```

Comment:
```suggestion
    if (false == res) {
        close_writer();
        return false;
    }

    m_is_open = true;

    // Flush the preamble
    flush();

    return true;
```
we should return `false` to indicate error right?